### PR TITLE
GUACAMOLE-793: Add support for retrieving effective groups from CAS.

### PR DESCRIPTION
**Pull request by mike-jumper**
This PR builds upon the changes proposed by @doctorfree via #563 (which itself built upon #454 from @siacali), aiming to clean up the LDAP parsing logic such that things are actually LDAP-aware. I did this by:

1. Pulling the changes from #563.
2. Squashing those changes down to a single commit, preserving authorship and the original commit message.
3. Building off of that.

Instead of regex-driven parsing of LDAP DNs, these changes leverage `LdapName` to truly parse each DN. As CAS may also expose multiple effective groups through a `Collection`, these changes also specifically test for this and include each group therein.

I have refactored the configuration properties to reflect this:

* `cas-group-attribute` - The CAS attribute that determines group membership. If omitted, groups aren't retrieved from CAS, and all other group-related properties for CAS are ignored.
* `cas-group-format` - The format that CAS will use for its group names. This may be `plain` (simple text names) or `ldap` (LDAP DNs). If set to `ldap`, group names are _always_ determined from the last (leftmost) attribute of the DN. If omitted, `plain` is used by default.
* `cas-group-ldap-base-dn` - The base DN to require for LDAP-formatted CAS groups. If specified, only CAS groups beneath this DN will be included. This property has no effect if `cas-group-format` is not `ldap`.
* `cas-group-ldap-attribute` - The LDAP attribute to require for LDAP-formatted CAS groups. If specified, only CAS groups that use this attribute for the name of the group will be included. This property has no effect if `cas-group-format` is not `ldap`.

Many thanks to ...

* @siacali for taking the first stabs at CAS group support.
* @doctorfree for further, repeated stabbing.
* @lchanouha for pointing toward `LdapName` as a good means of parsing an LDAP DN, and confirming that the CAS library (sometimes?) leverages `Collection` to store multi-valued attributes.

**Beware that I do not have a CAS server that I can test these changes against.** If anyone here _does_ (or knows how I might easily spin one up), it would be good to recheck that all this works as expected.

### CODE REVIEW

**Comment by lchanouha** [comment]
Many thanks, looks great to me 👍 
Will try it monday.

Could you add the 4 properties to guacamole-docker/bin/start.sh for Docker envs ?

**Review by necouchman — CHANGES_REQUESTED** [review]
A couple of comments, looks pretty good.

I have a CAS server and will test against that before merging.

File: extensions/guacamole-auth-cas/src/main/java/org/apache/guacamole/auth/cas/AuthenticationProviderService.java [hunk 1]
```diff
@@ -34,13 +34,20 @@
...
  * Service providing convenience functions for the CAS AuthenticationProvider
  * implementation.
  */
 public class AuthenticationProviderService {
 
+    /**
+     * Logger for this class.
+     */
+    private static final Logger logger = LoggerFactory.getLogger(AuthenticationProviderService.class);
+
```
[the remark this answers is not in the document]
Comment (mike-jumper) [reply extensions/guacamole-auth-cas/src/main/java/org/apache/guacamole/auth/cas/AuthenticationProviderService.java:50]: I don't believe so - I think this crept in from 7b8dc36644be90f5ddb90e4d7835e062475421ea. I'll recheck and remove.

File: extensions/guacamole-auth-cas/src/main/java/org/apache/guacamole/auth/cas/conf/ConfigurationService.java [hunk 2]
```diff
@@ -85,4 +91,102 @@ public PrivateKey getClearpassKey() throws GuacamoleException {
...
+     */
+    public GroupFormat getGroupFormat() throws GuacamoleException {
+        return environment.getProperty(CASGuacamoleProperties.CAS_GROUP_FORMAT, GroupFormat.PLAIN);
+    }
+
+    /**
+     * Returns the base DN that all LDAP-formatted CAS groups must reside
+     * beneath. Any groups that are not beneath this base DN should be ignored.
+     * If no such base DN is provided, the tree structure of the ancestors of
+     * LDAP-formatted CAS groups should not be considered.
```
[the remark this answers is not in the document]
Comment (mike-jumper) [reply extensions/guacamole-auth-cas/src/main/java/org/apache/guacamole/auth/cas/conf/ConfigurationService.java:130]: That all groups will be allowed.

For example, if `cas-group-format` is `ldap` and CAS returns the following groups:

* `cn=group1,ou=groups,dc=foo,dc=example,dc=net`
* `cn=group2,ou=groups,dc=bar,dc=example,dc=net`
* `cn=group3,ou=groups,dc=baz,dc=example,dc=net`

Then, by default, the user will be considered a member of "group1", "group2", and "group3". However, if `cas-group-ldap-base-dn` is set to `dc=bar,dc=example,dc=net`, they will only be considered a member of "group2".

**Comment by mike-jumper** [comment]
> ...
> Could you add the 4 properties to guacamole-docker/bin/start.sh for Docker envs ?

Sure - added via commit e6a2a08822834792ba91911acdbc3c91c98c6e41.

**Comment by lchanouha** [comment]
OK I successfully validated a working LDAP + CAS group matching.

I tested both cas-group-format=plain & cas-group-format=ldap, cas-group-ldap-base-dn, cas-group-ldap-attribute, cas-group-attribute.
Docker variables matching works too.

Many thanks.

**Comment by necouchman** [comment]
Thanks for testing @lchanouha! I've been trying to test, myself, but I'm having issues with any attribute release on my CAS server right now, so I think it's something with my CAS installation. I'll go ahead and approve/merge this request and still try to do some testing myself to see if I can get things working properly.

**Review by necouchman — APPROVED** [review]

**Comment by lchanouha** [comment]
OK. 
The easy way to setup a CAS server is the Docker version of apereo https://github.com/apereo/cas-overlay-template
You just need to setup a attribute repository in cas.properties:
```
cas.authn.accept.users=
cas.authn.ldap[0].type=ANONYMOUS
cas.authn.ldap[0].ldapUrl=ldaps://ldap1.fqdn.fr ldaps://ldap2.fqdn.fr
cas.authn.ldap[0].baseDn=ou=People,dc=fqdn,dc=fr
cas.authn.ldap[0].subtreeSearch=true
cas.authn.ldap[0].searchFilter=(uid={user})
cas.authn.ldap[0].principalAttributeId=uid
cas.authn.ldap[0].poolPassivator=NONE

cas.authn.attribute-repository.ldap[0].attributes.memberOf=memberOf
cas.authn.attribute-repository.ldap[0].ldapUrl=ldaps://ldap1.fqdn.fr ldaps://ldap2.fqdn.fr
cas.authn.attribute-repository.ldap[0].useStartTls=false
cas.authn.attribute-repository.ldap[0].baseDn=ou=People,dc=fqdn,dc=fr
cas.authn.attribute-repository.ldap[0].searchFilter=uid={0}
cas.authn.attribute-repository.ldap[0].poolPassivator=NONE
```

**Comment by necouchman** [comment]
I think the only difference I have is that, instead of the attribute repository, I'm trying to pull attributes directly with the LDAP authentication query:
    
    cas.authn.ldap[0].principalAttributeList=sn,cn,givenName,mail,memberOf

and then I have the JSON service configured similar to what you have:
```
{
    "@class" : "org.apereo.cas.services.RegexRegisteredService",
    "serviceId" : "^https://guacamole.example.com/.*",
    "name" : "guacamole HTTPS",
    "id" : 9999,
    "evaluationOrder": 9999,
    "attributeReleasePolicy" : {
        "@class" : "org.apereo.cas.services.ReturnAllAttributeReleasePolicy"
    }
}
```

But for some reason all of the attributes are coming back empty. I'll try the attribute repository route and see what happens.

**Comment by doctorfree** [comment]
Thanks! Looks good. Glad we worked through the process to arrive at a good solution.

**Pull request merged**